assembly/src/assembler/mod.rs

```diff
@@ -646,32 +679,111 @@ impl Assembler {
+    ///
+    /// If no [`MastNodeId`] exists for that procedure root, we wrap the root in an
+    /// [`crate::mast::ExternalNode`], and return the resulting [`MastNodeId`].
     pub(super) fn resolve_target(
         &self,
         kind: InvokeKind,
         target: &InvocationTarget,
         proc_ctx: &ProcedureContext,
-        mast_forest_builder: &MastForestBuilder,
-    ) -> Result<RpoDigest, AssemblyError> {
+        mast_forest_builder: &mut MastForestBuilder,
+    ) -> Result<MastNodeId, AssemblyError> {
         let caller = CallerInfo {
             span: target.span(),
             module: proc_ctx.id().module,
             kind,
         };
         let resolved = self.module_graph.resolve_target(&caller, target)?;
         match resolved {
-            ResolvedTarget::Phantom(digest) => Ok(digest),
+            ResolvedTarget::Phantom(mast_root) => {
+                self.ensure_valid_procedure_mast_root(
+                    kind,
+                    target.span(),
+                    mast_root,
+                    mast_forest_builder,
+                )?;
+
+                mast_forest_builder.ensure_external(mast_root)
+            },
             ResolvedTarget::Exact { gid } | ResolvedTarget::Resolved { gid, .. } => {
-                match mast_forest_builder.get_procedure_hash(gid) {
-                    Some(proc_hash) => Ok(proc_hash),
+                match mast_forest_builder.get_procedure(gid) {
+                    Some(proc) => Ok(proc.body_node_id()),
+                    // We didn't find the procedure in our current MAST forest. We still need to
+                    // check if it exists in one of a library dependency.
                     None => match self.module_graph.get_procedure_unsafe(gid) {
-                        ProcedureWrapper::Info(p) => Ok(p.digest),
-                        ProcedureWrapper::Ast(_) => panic!("Did not find procedure {gid:?} neither in module graph nor procedure cache"),
+                        ProcedureWrapper::Info(p) => {
+                            self.ensure_valid_procedure_mast_root(
+                            kind,
+                            target.span(),
+                            p.digest,
+                            mast_forest_builder,
+                        )?;
+
+                        mast_forest_builder.ensure_external(p.digest)
+                    },
+                        ProcedureWrapper::Ast(_) => panic!("AST procedure {gid:?} exits in the module graph but not in the MastForestBuilder"),
                     },
                 }
-            }
+            },
         }
     }
+
+    /// Verifies the validity of the MAST root as a procedure root.
+    fn ensure_valid_procedure_mast_root(
+        &self,
+        kind: InvokeKind,
+        span: SourceSpan,
+        mast_root: RpoDigest,
+        mast_forest_builder: &mut MastForestBuilder,
+    ) -> Result<(), AssemblyError> {
```

Comment:
I think the comment is a bit confusing here - i.e., what does it mean for the MAST root to be valid as a procedure root?

Also:
- If this function just verifies, maybe the name should be something like `verify_mast_root_is_procedure_root`?
- It doesn't seem like we need `mast_forest_builder` here to be mutable.

Also also: it seems like the way this function is used is that it is always followed by a call to `mast_forest_builder.ensure_external(p.digest)`. Maybe we can move this call into this function (and then the name `ensures` makes more sense and also `mast_forest_builder` can remain mutable).

Reply:
I moved the call to `ensure_external` back in and adjusted the docstring